sdk/trace.go

```diff
@@ -0,0 +1,123 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package sdk
+
+import (
+	"context"
+
+	"go.opentelemetry.io/collector/pdata/ptrace"
```

Comment:
why do we use the collector's objects here? Just curious

Reply:
That's a good question. I'm not sure we do in the long term and it's something I had mentioned in the outstanding issues of https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/1045.

tl;dr: I have added it primary as a quick way to get this off the ground, and this is an already working binary encoding.

We need some encoding to transmit span data. As @RonFed  pointed out in a recent SIG meeting, trying to encode all the Go specific span data in C is challenging. Given we are going to have a processing chain of `Go -> eBPF (C) -> Go` if we have an encoding that is understood on both "Go" ends and can be transmitted though some generic `byte` structure in C we can avoid a lot of complexity.

The collector `ptrace` is an already highly-optimized binary encoding that provides a Go API. It is based off of OTLP and is designed to be highly performant. That means that we can start with this and get a working concept.

However, in the long run, there is still an open question as to whether we want to use this. The SDK is not included in the published module sets and it is something we should decide on before doing so.

Answering if we want to replace `ptrace` with our own binary encoding opens up some interesting questions:

- The collector is not stable, can we rely on `ptraces` if we want to go stable?
- Can we borrow tooling from the collector to make our own?
- The `Traces` type is over-sized in that it holds multiple spans. We only transmit one span at a time currently, do we need all the overhead of `Traces`? Should we update to utilize the full scope of `Traces` (see next).
- Should we fully pivot to using `pdata` as our data model? Instead of interfacing with an opentelemetry-go SDK should we interface with the collector receiver interface directly[^1]?
  - This means our public `Interface` type would be an instance of the collector instead of wrapping the opentelemetry-go APIs

[^1]: https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1026

Reply:
> Should we fully pivot to using `pdata` as our data model? Instead of interfacing with an opentelemetry-go SDK should we interface with the collector receiver interface directly[1](#user-content-fn-1-e6622de7b09db4912a65bf020341a161)?
> * This means our public `Instrumentation` type would be an instance of the collector instead of wrapping the opentelemetry-go APIs

This would mean, that instead of having something like this:

https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/1025/files#diff-f55965a613ff264fe2e580d1779471172a8c8a906c321256630aa0fdfed8eef3R13-R16

We would just use `ptrace.Traces` as the argument to `Controller.Trace`.